From 3e6cc50ba404cb774d5d0c246dc6b320b3b484c0 Mon Sep 17 00:00:00 2001
From: Kent Tamura <tkent@chromium.org>
Date: Tue, 21 May 2024 05:48:47 +0000
Subject: [PATCH] Ruby: 'display: inline flow list-item` should inlinify children

... as `display: inline flow` does.
This CL fixes a crash by line-breakable ruby.

Bug: 340918183
Change-Id: Ic27b3e34e23fd7a6008802c907f206332774d680
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5552850
Commit-Queue: Koji Ishii <kojii@chromium.org>
Auto-Submit: Kent Tamura <tkent@chromium.org>
Reviewed-by: Koji Ishii <kojii@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1303609}
---

--- a/third_party/blink/renderer/core/style/computed_style.h
+++ b/third_party/blink/renderer/core/style/computed_style.h
@@ -1659,7 +1659,8 @@
     // If an inline box (inline flow) is inlinified, it recursively inlinifies
     // all of its in-flow children
     return IsInInlinifyingDisplay() &&
-           (display == EDisplay::kContents || display == EDisplay::kInline);
+           (display == EDisplay::kContents || display == EDisplay::kInline ||
+            display == EDisplay::kInlineListItem);
   }
 
   // Return true if an element with this computed style requires LayoutNG
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-ruby/ruby-inlinize-recursive-simple-ref.html
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
+<meta charset="utf-8">
+<style>
+li { display: inline-block; }
+rt { display: ruby-text; }
+</style>
+<body>
+
+<div><ruby><b>firstsecond</b><li>firstsecond</li></ruby></div>
+<div><rt><b>firstsecond</b><li>firstsecond</li></rt></div>
+
+</body>
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-ruby/ruby-inlinize-recursive-simple.html
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
+<meta charset="utf-8">
+<title>Inlinize block-level boxes recursively inside ruby and rt</title>
+<link rel="help" href="https://drafts.csswg.org/css-ruby/#anon-gen-inlinize">
+<link rel="help" href="https://drafts.csswg.org/css-display-4/#inlinify">
+<link rel="match" href="ruby-inlinize-recursive-simple-ref.html">
+<style>
+li { display: inline flow list-item; list-style-type: none; }
+.bl { display: block; }
+rt { display: ruby-text; }
+</style>
+<body>
+
+<div><ruby><b><span class="bl">first</span><span class="bl">second</span></b><li><span class="bl">first</span><span class="bl">second</span></li></ruby></div>
+<div><rt><b><span class="bl">first</span><span class="bl">second</span></b><li><span class="bl">first</span><span class="bl">second</span></li></rt></div>
+
+</body>
